### PR TITLE
Improvement: improvements to caching of attributes

### PR DESCRIPTION
**Pull request by ilya-causalens**
## Description and Motivation

- Improved caching of results for `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph` class, including:
  - Adding caching to the following methods: `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.get_minimal_graph`, `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.is_minimal_graph`,
    `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.is_stationary_graph`.
  - Fixed a bug in `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.get_stationary_graph` where the same cached graph would be returned multiple times. Instead now,
    a deepcopied version of this graph is returned.
  - Fixed a bug where cached attributes would not be reset when adding edges to `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph`, which can lead to erroneous
    stationary graphs being returned.
- Added caching of `cai_causal_graph.causal_graph.CausalGraph.is_dag`, `cai_causal_graph.causal_graph.CausalGraph.adjacency_matrix` and `cai_causal_graph.causal_graph.CausalGraph.to_networkx`
  methods.
- Unified cached attribute resetting between `cai_causal_graph.causal_graph.CausalGraph` and `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph`
  by introducing a `cai_causal_graph.causal_graph.CausalGraph._reset_cached_attributes` method and `cai_causal_graph.causal_graph.reset_cached_attributes_decorator`
  decorator.
- Fixed a bug where calling `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.variables` method would return the variable list by reference, rather than returning
  a copy.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [ ] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [ ] I have implemented all requirements? (see JIRA, project documentation)
- [ ] I am affecting someone else's work? If so: I have added those people as reviewers?
- [ ] I have added comments to all the bits that are hard to follow
- [ ] At a bare minimum, I tested this manually
- [ ] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [ ] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [ ] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

File: cai_causal_graph/causal_graph.py [hunk 1]
```diff
@@ -614,6 +633,8 @@ def get_identifier(self) -> str:
 
     def is_dag(self) -> bool:
         """Check whether the `cai_causal_graph.causal_graph.CausalGraph` instance is a Directed Acyclic Graph (DAG)."""
+        if self._is_dag is not None:
+            return self._is_dag
         return networkx.is_directed_acyclic_graph(self.to_networkx()) if self._is_fully_directed() else False
```
Comment (andrew-causalens) [inline cai_causal_graph/causal_graph.py:638]: Should we cache `self._is_fully_directed()`, `self._is_fully_undirected()`, and `self.to_networkx()` as well? Could actually cache other `to_` methods but think it is overkill. The networkx one may be useful though.

**Review by andrew-causalens — COMMENTED** [review]
Looks great and huge improvement! Don't think `_is_dag` is being cached though.

File: cai_causal_graph/causal_graph.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilya-causalens) [reply cai_causal_graph/causal_graph.py:638]: Ah good catch. Fixed.

Comment (ilya-causalens) [reply cai_causal_graph/causal_graph.py:638]: Done - cached `to_networkx` and `adjacency_matrix` too.

**Review by andrew-causalens — APPROVED** [review]

**Pull request merged**